JTAG support for CMSIS-DAP (take 2)

cstrahan opened the pull request:
This builds on @prabhpreet's work in #1216.

This PR needs some polish, and I could use some advice on how to structure some things (I fully intend to clean up any `todo!()`s and such).

Here's where it stands as of writing:

* I can successfully flash and debug my Black Pill with this branch.
* ... granted, I currently hard-code the IR lengths to 4 and 5 for device 0 and 1, respectively.
* However, the functions to scan for connected JTAG devices works -- I just need some advice on where this should be performed, and how to plumb that into the place where I ultimately need to send the `DAP_JTAG_Configure` request.

I might be able to come up with some ideas for that last bullet point on my own by looking at how the FTDI probe works / is configured (as I know it has a scanning routine as well)... but it's late and I need to get to bed now :).

I would love to get this in a state worth merging, so gimme direction and let's get this over the finish line :).

cstrahan commented:
So, I'm really not happy with this sort of thing:

```rust
const NO_CAPTURE: bool = false;
const TDI_ZEROES: [u8; 8] = [0x00; 8];
const TMS_LOW: bool = false;
JtagSequence::new(1, NO_CAPTURE, TMS_LOW, TDI_ZEROES)?,
```

Curious if I should just make the `JtagSequence` fields public and remove the `fn new`, or some other approach? :thinking: 

Without the constants, it's kinda hard to read, for instance:

```rust
JtagSequence::new(1, false, false, 0u64.to_le_bytes())?,
```

Also, while the SWJ Sequence type uses a byte array... I kinda wonder if it would be cleaner over all to just use a `u64`.

Open to suggestions.

cstrahan commented:
The ftdi code does this when attaching:

```rust
    fn attach(&mut self) -> Result<(), DebugProbeError> {
        tracing::debug!("attaching...");

        self.adapter
            .attach()
            .map_err(|e| DebugProbeError::ProbeSpecific(Box::new(e)))?;

        let taps = self
            .adapter
            .scan()
            .map_err(|e| DebugProbeError::ProbeSpecific(Box::new(e)))?;
        if taps.is_empty() {
            tracing::warn!("no JTAG taps detected");
            return Err(DebugProbeError::TargetNotFound);
        }
        if taps.len() == 1 {
            self.adapter
                .select_target(taps[0].idcode)
                .map_err(|e| DebugProbeError::ProbeSpecific(Box::new(e)))?;
        } else {
            let known_idcodes = [
                0x1000563d, // GD32VF103
            ];
            let idcode = taps
                .iter()
                .map(|tap| tap.idcode)
                .find(|idcode| known_idcodes.iter().any(|v| v == idcode));
            if let Some(idcode) = idcode {
                self.adapter
                    .select_target(idcode)
                    .map_err(|e| DebugProbeError::ProbeSpecific(Box::new(e)))?;
            } else {
                return Err(DebugProbeError::TargetNotFound);
            }
        }
        Ok(())
    }
```

So if there's more than one TAP, it currently looks for a GD32VF103 (idcode: 0x1000563d) and uses that; otherwise it fails.

If we want to get this in sooner than later, I suppose we could do something similar -- I could do something similar for the STM32F4X (and related families) so that at least they would work out-of-the-box without introducing any new user config changes.

If you want me to take a stab at amending the configuration as necessary to expose these details for JTAG setups, lemme know. (I'll probably poke around at it any way, but I'm split on whether that should be part of this PR's scope).

cstrahan commented:
So where do we want to perform the JTAG scanning and configuration?

Currently, the switch from SWD-to-JTAG (or vice versa) is performed in `ArmDebugSequence::debug_port_setup`:

https://github.com/probe-rs/probe-rs/blob/ab03d7445b551354dbe6d913882759e4c07a87f2/probe-rs/src/architecture/arm/sequences/mod.rs#L447-L492

The problem here (as I see it) is that this inverts the control from the way we probably want it. In order to enumerate JTAG TAPs/DAPs, the device(s) need to first be configured to use JTAG (via SWJ-DP Switch, if necessary). `ArmDebugSequence` seems like the wrong place for this; ideally the switching should be trivially performed given one of the lower level debug probe interfaces like `DebugProbe`. Actually, if you look at `CmsisDap::attach`, you'll find that's where we currently configured SWD:

https://github.com/probe-rs/probe-rs/blob/ab03d7445b551354dbe6d913882759e4c07a87f2/probe-rs/src/probe/cmsisdap/mod.rs#L452-L474

So it would seem that configuring the CMSIS-DAP probe's JTAG settings should also happen there. Assuming the pattern set by the `attach` in the ftdi module is something we like, I think we should do something like:

* Perform the SWJ-DP switch in `attach`
* Assuming the user hasn't explicitly declared the JTAG chain details in their config (at which index the device is, and each of the IR lengths), then scan for devices:
  * If there's just one, we can use it without further thought (although, if the user has specified an expected IDCODE, we could return here if a check for the IDCODE doesn't match)
  * If there are multiple devices, we try to scan for IR lengths and IDCODEs; we'll require that the user specify at least an expected IDCODE for the target so we can determine the target's index (or as another option, they could just specify the index into the JTAG chain directly).

I'm not sure how we'd want to plumb the IDCODE into the `CmsisDap` though. Maybe we add a new function to `DebugProbe` like `set_jtag_chain`, and you'd call that with the necessary info prior to calling `attach`?

cstrahan commented:
Hey @noppej, since you touched target-gen recently, I was curious if you might have some thoughts on where (and maybe part of _how_) to specify a given chip's TAP(s) info. For instance, the STM32F401x I'm working with has two TAPs wired serially within the same chip (one for boundary scan, the other for ARM debugging -- the former being disabled when SWJ-DP is switched into SWD mode):

>#### 23.5 STM32F401xB/C and STM32F401xD/E JTAG TAP connection
>The STM32F401xB/C and STM32F401xD/E MCUs integrate two serially connected JTAG TAPs, the boundary scan TAP (IR is 5-bit wide) and the Cortex ®-M4 with FPU TAP (IR is 4-bit wide).
>
>To access the TAP of the Cortex ®-M4 with FPU for debug purposes:
>
>1. First, it is necessary to shift the BYPASS instruction of the boundary scan TAP.
>2. Then, for each IR shift, the scan chain contains 9 bits (=5+4) and the unused TAP instruction must be shifted in using the BYPASS instruction.
>3. For each data shift, the unused TAP, which is in BYPASS mode, adds 1 extra data bit in the data scan chain.
>
> Note: Important: Once Serial-Wire is selected using the dedicated Arm® JTAG sequence, the
boundary scan TAP is automatically disabled (JTMS forced high).

I believe we ought to include this info (including IR length of each respective TAP) in `probe-rs` so the user can specify the target chip and have the JTAG chain thereby inferred (in the happy path where _only_ that chip is on the chain). Do you have any thoughts on whether this data could/should be stored in the existing target descriptions, or in an ancillary file? Do you know if the CMSIS pack files include anything about this?

noppej commented:
@cstrahan Thanks for asking, but this is probably a bit above my pay grade :)  @Yatekii / @Tiwalun ... any suggestions here?

cstrahan commented:
@noppej Haha, yeah, I didn't know about any of this until a couple months ago, so I think it's outside my pay grade as well :laughing:. Thanks for your quick response though!

I downloaded/unpacked one of the CMSIS packs and grepped for the expected IDCODEs, and nothing turned up -- so my informed guess would be that the pack format doesn't capture these details. Of course, I could read through the 1 million pages of the CMSIS Pack spec to get a definitive answer, but... nah, I'mma pass on that.

Tiwalun commented:
> I downloaded/unpacked one of the CMSIS packs and grepped for the expected IDCODEs, and nothing turned up -- so my informed guess would be that the pack format doesn't capture these details. Of course, I could read through the 1 million pages of the CMSIS Pack spec to get a definitive answer, but... nah, I'mma pass on that.

The pack specification has an entry for JTAG information: 

```xml
<family Dfamily="STM32F4" Dvendor="STMicroelectronics:13">
  ...
  <debugconfig default="swd" clock="10000000" swj="1"/>
  
  <debugport __dp="0">
    <jtag tapindex="0" idcode="0x4BA00477" irlen="4"/>
    <swd  idcode="0x2BA01477"/>
  </debugport>
  ...
</family>
```

See here: https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/pdsc_family_pg.html#element_dp_jtag

Unfortunately that doesn't necessarily mean that anybody uses it.

noppej commented:
That's promising. If I understand it correctly, then that means we can target-gen it and use that information as part of the target definition. Right?

cstrahan commented:
@Tiwalun Awesome -- thanks for the reference! Unfortunate (and ironic, given the example) that not all manufacturers care to specify this :disappointed:.

@noppej For packs that specify these (unlike `Keil.STM32F4xx_DFP.pdsc`), it looks like we could at least get the expected IDCODEs for SWD/JTAG, and the IR length. In the case where there are multiple TAPs wired serially, I'm still unsure if well behaved packs (or more importantly, the typical packs) will care to include the non-ARM debug TAPs (e.g. boundary scan) -- then again, I suspect that represents a small minority of ARM devices that even have multiple TAPs.

Is there a convenient way to download all of the packs for which we have already generated target  YAMLs? I would be interested in seeing how common these configs are among the chips we already support.

noppej commented:
> Is there a convenient way to download all of the packs for which we have already generated target YAMLs? I would be interested in seeing how common these configs are among the chips we already support.

`target-gen arm --help` ... by default it pulls all the packs we already supports, but you can add a `--filter` to reduce scope.

Tiwalun commented:
> That's promising. If I understand it correctly, then that means we can target-gen it and use that information as part of the target definition. Right?

Technically yes. But for example the pack for the STM32F401 mentioned earlier doesn't contain it. In a first step I wouldn't bother too much with the target definition part, we already have a hardcoded implementation for the FTDI probe... I'm totally fine with cleaning that up in a later PR, and just hardcoding an IDCODE here as well.

That will also help to keep the PR a bit smaller.


> The problem here (as I see it) is that this inverts the control from the way we probably want it. In order to enumerate JTAG TAPs/DAPs, the device(s) need to first be configured to use JTAG (via SWJ-DP Switch, if necessary). `ArmDebugSequence` seems like the wrong place for this; ideally the switching should be trivially performed given one of the lower level debug probe interfaces like `DebugProbe`.

The implementation of these sequences is based on the CMSIS-Packs, so the function is a copy of the default sequence from there. The idea behind this is to be (one day) able to import sequences from the packs automatically.

The [original sequence in the spec](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/debug_description.html#debugPortSetup) also includes the JTAG code. I'd like to keep it that way in probe-rs as well for now, so it would be better to extend it there. I don't recall why we do it again in the `attach` function, to be honest.

cstrahan commented:
@noppej Ah, it wasn't clear to me that it would download everything by default (I was actually going to try specifying an empty name prefix filter like `-f ''` to see if that would do the trick, haha). Thanks!

cstrahan commented:
@Tiwalun

> The implementation of these sequences is based on the CMSIS-Packs, so the function is a copy of the default sequence from there. The idea behind this is to be (one day) able to import sequences from the packs automatically.

That rationale make sense -- thanks!

> The [original sequence in the spec](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/debug_description.html#debugPortSetup) also includes the JTAG code.

It has the code to transfer properly switch to JTAG, reset and enter Run-Test/Idle (which I referenced to make my changes in `debug_port_setup`, but that still leaves the question of where/when to tell the attached CMSIS-DAP probe about the JTAG chain via DAP_JTAG_Configure. Should we do this in `debug_port_setup`, say immediately after performing the sequence from the spec? That would also entail performing any requisite auto scanning at that point.

> we already have a hardcoded implementation for the FTDI probe... I'm totally fine with cleaning that up in a later PR, and just hardcoding an IDCODE here as well

I think this is a fine strategy. I can commit to working on a more flexible solution after we get this PR in.

MabezDev commented:
@cstrahan Thanks for your work on this! Now that J-LINK supports CMSIS-DAP I'd love to get this support added in, the default driver is painfully slow. What's the status of this PR?

cstrahan commented:
@MabezDev It essentially works -- I've been using it for all of my flashing/debugging needs with my Teensy MicroMod (with hardware modification to break out the JTAG signals). Unfortunately, I've had to almost entirely suspend this work (along with all of my other projects) while I look for a new job (got laid off at the start of the year).

Here's what should/needs to change:

- Some clean up (I requested suggestions to clean up some of my concerns -- or at the very least bless what I have -- but I haven't heard back yet). So I'll need some time to do some head scratching and settle on how to refactor the things that irk me.
- A sane, if not complete (for now), approach to handling multiple DAPs on the JTAG chain. My working tree has a small hack to work with one of my STM devices, which has two DAPs chained internally in the ARM chip (which means lack of support for multiple DAPs implies probe-rs won't work for this type of device).

I think that's all that remains.

I really would love to get this wrapped up, though -- both so everyone can benefit from it, as well as to avoid any divergence from master. I'll see if I can squeeze in a few hours at some point, but that's gonna be tricky.

cstrahan commented:
@MabezDev Also, FWIW, I've been enjoying my MCU-Link probe (stopped using my J-Link), using the latest stock firmware. You can pick one up for about $15.

(I also tried the DAPLink firmware on it, but I [ran into some problems there](https://github.com/ARMmbed/DAPLink/issues/1004) -- a shame, because DAPLink supports nifty things in addition to exposing a CMSIS-DAP interface, like flashing over a mass storage interface, which I believe should be much faster than speeds achievable over CMSIS-DAP itself).

cstrahan commented:
Latest commit (24e6ff52a1469eea53acb2546cee4dc94a392e97):

1. replaces my scanning algorithm with [code adapted from `jtagdap`](https://github.com/adamgreig/jtagdap/blob/19dcb6af34f0f9843cba3cecaf3eb2a53b2d5a55/src/jtag.rs) (jtagdap is dual licensed under MIT and Apache).
2. Removes all warnings.
3. `debug_port_setup` now calls `interface.configure_jtag()`, which (as part of the scanning) will attempt to discover IR lengths. Later on, we can figure out how we want to permit the user to configure expected IR lengths.

I still want to clean up the code surrounding `JtagSequence::new`, as I described earlier. And maybe introduce a new error type for IDCODE issues.

Apart from cleanliness, and anything I may have missed (do let me know!), I think this is functionally complete.

cstrahan commented:
> I still want to clean up the code surrounding `JtagSequence::new`, as I described earlier. And maybe introduce a new error type for IDCODE issues.

This is done in https://github.com/probe-rs/probe-rs/pull/1462/commits/cf8f5da9bd897499323836b886152c3da8343901

Now there are `Sequence::capture` and `Sequence::no_capture` constructor functions.

An example of a sequence to enter Test-Logic-Reset (which requires six clock cycles with TMS high and TDI low):

```rust
let sequence = JtagSequence::no_capture(true, &bitvec![u8, Lsb0; 0; 6])?;
```

The number of clock cycles is taken from the bitvec argument, which in turn consists of TDI bits.

I think that's a lot easier to read and write than:

```rust
let sequence = JtagSequence::new(6, false, true, 064.to_le_bytes())?;
```

---

With that, I think I've done all I really want to do with this PR.

This is ready for a proper review.

cstrahan commented:
The JTAG scan chain detection in action, from my logs:

```
{"timestamp":"2023-07-02T09:08:16.397833Z","level":"DEBUG","fields":{"message":"JTAG DR scan chain detected as 32 bits long"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":502,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-02T09:08:16.398648Z","level":"DEBUG","fields":{"message":"JTAG IR scan chain detected as 4 bits long"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":502,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-02T09:08:16.398846Z","level":"DEBUG","fields":{"message":"Found IDCODE: 0x0BA02477 (ARM Ltd)"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":284,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-02T09:08:16.398910Z","level":"INFO","fields":{"message":"Only one TAP detected, IR length 4"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":387,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
```

(Is there a recommended/go-to log file filter to print these traces in a more human friendly way? :thinking:)

cstrahan commented:
To aid the reviewing process, here's a description of how the scan chain interrogation works:

First we find all IDCODES.

This works by entering Test-Logic-Reset, which initializes the instruction register of each TAP to "contain the IDCODE instruction or, if the optional device identification register is not provided, the BYPASS instruction"[^1]

[^1]: [IEEE-Std-1149.1-2001](https://fiona.dmcs.pl/~cmaj/JTAG/JTAG_IEEE-Std-1149.1-2001.pdf) section 6.1.2, "The TAP Controller"

Now we enter Capture-DR.

We shift-in 0s into DR, and capture the shifted-out TDO. For each TAP, the captured TDO contains either the respective IDCODE (if the instruction is supported), or a single 0 (in the case of BYPASS). Call this captured TDO `d0_dr`.

We shift-in 1s into DR, and capture the shifted-out TDO. By finding the index of the first 1, we can determine the length of the DR scan chain. Call this length `dr_length`.

We can now truncate `d0_dr` to `d0_dr[0..dr_length]`.

We can now drain IDCODEs out of `d0_dr` by checking the first bit[^2]: if it is a 1 then we know it is a IDCODE and can drain 32 bits; it is a 0 then we know that this is a single 0 bit from a BYPASS instruction.

[^2]:  [IEEE-Std-1149.1-2001](https://fiona.dmcs.pl/~cmaj/JTAG/JTAG_IEEE-Std-1149.1-2001.pdf)  section 12.1 "Design and operation of the device identification register":  *Since the bypass register (which is selected in the absence of a device identification register by the instruction loaded in the Test-Logic-Reset controller state) loads a logic 0 at the start of a scan cycle, whereas a device identification register will load a constant logic 1 into its LSB, examination of the first bit of data shifted out of a component during a test data scan sequence immediately after exit from the Test-Logic-Reset controller state will show whether a device identification register is included in the design.*

Now we know how many TAPs are on the scan chain, and for those that support IDCODE instruction, we have their respective IDCODE.

Now we can find the IR scan chain length[^3]. This proceeds similar to what we did to collect the IDCODEs.

[^3]: We need the IR lengths of each TAP so we can inform the probe. The probe needs this info because it will will need to know where in the chain our target TAP is, and also to set all other taps before/after it into BYPASS.

Now we enter Capture-IR.

We shift-in 0s into IR, and capture the shifted-out TDO. Call this captured TDO `d0_ir`.

We shift-in 1s into IR, and capture the shifted-out TDO. By finding the index of the first 1, we can determine the length of the IR scan chain. Call this length `ir_length`.

We can now truncate `d0_ir` to `d0_ir[0..ir_length]`.

We can now find _possible_ starting locations for each TAP's IR register within the IR scan chain by finding all indices within `d0_ir` where the bit pattern is `10`[^4]. If the number of possible starts is the same as the number of TAPs found earlier, we can unambiguously determine the starting location of each TAP within the IR scan chain; otherwise the user would need to specify the expected IR lengths and then we can at least sanity check those expected IR lengths against the list of possible starting locations. If the number of TAPs is less than the number of possible starts, we know the scan chain is broken somehow and we error out.

[^4]:  [IEEE-Std-1149.1-2001](https://fiona.dmcs.pl/~cmaj/JTAG/JTAG_IEEE-Std-1149.1-2001.pdf)  section 7.1.1 "Design and construction of the instruction register":  *The two least significant instruction register cells (i.e., those nearest the serial output) shall load a fixed binary “01” pattern (the 1 into the least significant bit location) in the Capture-IR controller state (see 7.2).* (note that the rest of the bits are unspecified, which is why these locations within the scan chain are only _possible_ starts for each TAP's IR register)

---

Some notes regarding implementation details:

When scanning DR/IR, I chose to shift-into the chain 128 bits per [`DAP_JTAG_Sequence`](https://arm-software.github.io/CMSIS_5/DAP/html/group__DAP__JTAG__Sequence.html) request (using two sequence infos each specifying 64 bits of TDI data). The max length we attempt to interrogate is set by:

```rust
// Max scan chain length (in bits) to attempt to detect.
const MAX_LENGTH: usize = 128;
```

Which currently is the same as the number of bits we shift in the aforementioned `DAP_JTAG_Sequence` request, which means we perform just one request to shift all the 0s, and another request to shift all the 1s. I structured the code this was because, at a later time, we aught to replace the `MAX_LENGTH` constant with a config parameter the user can specify. For the sake of efficiency, we would still shift 128 bits at a time, but this would allow the user to specify a larger upper bound on scan chain length.

cstrahan commented:
@seanmlyons22 

> +1 for this PR
> 
> I'm working on something adjacent and would benefit from your work here. Let me know if you would like to collaborate.

Cool! Can I ask what it is that you are working on?

cstrahan commented:
@Yatekii @Tiwalun 

If either of you want to review this together, I can make myself available for a meeting.

seanmlyons22 commented:
This is great work. I am looking into something similar/adjacent that could greatly benefit from what you have done here.

Let me know if there is anything that can be done to pitch in or help.

> @seanmlyons22
> 
> > +1 for this PR
> > I'm working on something adjacent and would benefit from your work here. Let me know if you would like to collaborate.
> 
> Cool! Can I ask what it is that you are working on?

I am looking at adding support for the TI CC family of chips to probe-rs (and rust in general). Started with the PAC and moving up. So right now I am working on your branch and trying to implement the cJTAG --> JTAG switch before leveraging your code for the JTAG part.

Yatekii commented on probe-rs/src/probe/cmsisdap/mod.rs:
```diff
@@ -520,7 +928,15 @@ impl DebugProbe for CmsisDap {
-        self.configure_swd(swd::configure::ConfigureRequest {})?;
+        if self.active_protocol() == Some(WireProtocol::Jtag) {
+            // no-op: we configure JTAG in debug_port_setup,
+            // becuase that is where we execute the SWJ-DP Switch Sequence
```
```suggestion
            // because that is where we execute the SWJ-DP Switch Sequence
```

Yatekii commented on probe-rs/src/probe/cmsisdap/mod.rs:
```diff
@@ -153,6 +221,346 @@ impl CmsisDap {
+            .map_err(CmsisDapError::from)
+            .and_then(|v| match v {
+                JtagSequenceResponse(Status::DAPOk, tdo) => Ok(tdo),
+                JtagSequenceResponse(Status::DAPError, _) => Err(CmsisDapError::ErrorResponse),
```
Can we print additional info here?

cstrahan replied:
@Yatekii  What extra info would you like to see here?

cstrahan replied:
Just to be clear: would you like to add another error variant (something like `JtagSequenceError`, maybe) that passes along the error code and sequence info up the call stack (to be printed however these errors are already printed), and/or do you suppose we should add a `trace::error` call right here?

One tricky thing either way (at least to my mind; let me know if there's an obvious solution I'm missing) is that a single sequence request can be represented as multiple sub sequences (you need a sequence per change to TMS, and sequences longer than 64 tck cycles also must be split up) -- so I don't know if we just dump the debug format and call it a day, or if we try to construct a more readily grokable textual representation of the sequence, like (spitballing here):

```
Failed to send 5 cycle JTAG sequence; requested logic levels:
    TMS: 11000
    TDI: 01010
```

(where I'm flattening the two sequences implied by the above -- there being at least two sequence because there's the transition from TMS high to TMS low)

I get the motivation to provide more info here in some way, though I'm not sure how to proceed.

Also, one other thing: the Status code is always either `0x00` for success or `0xFF` (excepting invalid response due to buggy probes and/or signal integrity problems), so I don't know if there's much practical utility in surfacing the status when there's an error.

```rust
#[derive(Copy, Clone, Debug)]
pub(crate) enum Status {
    DAPOk = 0x00,
    DAPError = 0xFF,
}

impl Status {
    pub fn from_byte(value: u8) -> Result<Self, SendError> {
        match value {
            0x00 => Ok(Status::DAPOk),
            0xFF => Ok(Status::DAPError),
            _ => Err(SendError::InvalidResponseStatus),
        }
    }
}
```

Yatekii replied:
Hmm, you are correct, it's not easily obvious what the correct action is here. I just thought of a simple `trace::debug!() ` call.
Maybe it will never be needed. We can revisit this when we hit an issue. I opened an issue in the meanwhile tracking this. But we can leave it as is. Thanks a lot!

Yatekii reviewed (commented):
This is looking very nice, thank you very much :)

I added some nits here and there and maybe @Tiwalun also finds time to look over it :)

cstrahan commented on probe-rs/src/architecture/arm/sequences/mod.rs:
```diff
@@ -466,11 +466,23 @@ pub trait ArmDebugSequence: Send + Sync {
                 // Change if SWJ-DP uses deprecated switch code (0xEDB6).
                 interface.swj_sequence(16, 0xE79E)?;
+
+                interface.swj_sequence(51, 0x0007_FFFF_FFFF_FFFF)?; // > 50 cycles SWDIO/TMS High.
```
I'll move the comments above :+1:

cstrahan commented on probe-rs/src/probe/jlink/arm.rs:
```diff
@@ -1354,6 +1354,32 @@ impl<Probe: DebugProbe + RawProtocolIo + JTAGAccess + 'static> RawDapAccess for
+        tms: bool,
+        mut bits: u64,
+    ) -> Result<(), DebugProbeError> {
+        let protocol = self.active_protocol().expect("No protocol set");
```
I followed the example set by `swj_sequence` below. Will have to think about that some more.

cstrahan replied:
For now, I've removed the expect in `jtag_sequence`. Not sure if we we should also do the same for `swj_sequence` :thinking:

Yatekii replied:
Actually this is not a public function. So we should make sure internally that this never panics. I think the `expect` was correct :/ I am sorry :(

cstrahan replied:
Done in https://github.com/probe-rs/probe-rs/pull/1462/commits/7cfc43290bb8c27b5f70db05c90fdf2b1f493a81

cstrahan commented on probe-rs/src/probe/jlink/arm.rs:
```diff
@@ -1354,6 +1354,32 @@ impl<Probe: DebugProbe + RawProtocolIo + JTAGAccess + 'static> RawDapAccess for
+        );
+
+        let mut io_bits = Vec::with_capacity(bit_len as usize);
+        for _ in 0..bit_len {
```
I modeled this after `swj_sequence` also. jaylink's `BitIter::new([...])` isn't public, unfortunately. 

But our espusbjtag does have its own `BitIter` -- I'll see if I can use that. However, `crate::probe::espusbjtag::protocol` is not `pub` (or `pub(crate)`). Have a recommendation for where I could factor BitIter out to?

Yatekii replied:
That's fair :) Maybe we can have a toplevel module for structures? I don't have a specific name preference.

cstrahan replied:
I created a `pub(crate)` module (unimaginatively) named `common` and moved `BitIter` there in this commit: https://github.com/probe-rs/probe-rs/pull/1462/commits/b13c7f8e8facfc94d55329198006e01c3ce0d99d

Yatekii replied:
Perfect! Thanks!

cstrahan commented on probe-rs/src/probe/cmsisdap/mod.rs:
```diff
@@ -65,6 +76,63 @@ pub struct CmsisDap {
+    }
+}
+
+impl IDCODE {
```
Latest commit does that :+1:

cstrahan commented:
Thanks @Yatekii for the review! I pushed some of the fixes. Gotta circle back when I'm free tomorrow.

cstrahan commented:
@Yatekii With few exceptions (where I've asked for clarification on how to proceed), I've made the changes you requested.

As of the last commit, I've tested changes by debugging my Teensy MicroMod (which *only* supports JTAG debugging, if you're willing to break out the signal lines) through VSCode with:
- My [MCU-LINK](https://www.nxp.com/design/software/development-software/mcuxpresso-software-and-tools-/mcu-link-debug-probe:MCU-LINK) (a CMSIS-DAP compliant probe), and
- My [Segger J-Link Base](https://www.segger.com/products/debug-probes/j-link/models/j-link-base/), to ensure the new `jtag_sequence` code path doesn't break `probe-rs` support for J-Link devices.

9names commented:
tested this with probe-rs info and benchmark - works great on my TI chips.
barfs on lpc4337, maybe a multicore issue? no idea.

trace and backtrace gist here: https://gist.github.com/9names/846b2cafc6a36acbb6c5252d9ae10671

Yatekii commented:
Thanks a lot! I added some clarifications where asked for :) I hope that helps :) In any case, it's all rather nitpicky than required changes.

cstrahan commented:
@9names Thank you for testing this and providing the trace!

The problem is here: https://github.com/cstrahan/probe-rs/blob/e76c2f9ed1f5c4554798515726e47ddf2c9eef02/probe-rs/src/probe/cmsisdap/mod.rs#L274

(on the `let idcode = dr.load::<u32>();` line)

```rust
    fn extract_idcodes(mut dr: &BitSlice<u8>) -> Result<Vec<Option<IdCode>>, CmsisDapError> {
        let mut idcodes = Vec::new();

        while !dr.is_empty() {
            if dr[0] {
                if dr.len() < 32 {
                    tracing::error!("Truncated IDCODE: {dr:02X?}");
                    return Err(CmsisDapError::InvalidIdCode);
                }

                let idcode = dr.load::<u32>();
                let idcode = IdCode(idcode);

                if !idcode.valid() {
                    tracing::error!("Invalid IDCODE: {:08X}", idcode.0);
                    return Err(CmsisDapError::InvalidIdCode);
                }
                tracing::debug!("Found IDCODE: {idcode}");
                idcodes.push(Some(idcode));
                dr = &dr[32..];
            } else {
                idcodes.push(None);
                tracing::debug!("Found bypass TAP");
                dr = &dr[1..];
            }
        }
        Ok(idcodes)
    }
```

The problem is in my use of `bitvec`'s `load`; per the docs:

> [Parameters](https://docs.rs/bitvec/latest/bitvec/field/trait.BitField.html#parameters-4)
>  - `&self`: A bit-slice region whose length is in the range `1 ..= I::BITS`.

The device I tested with has only one TAP, and the DR scan chain length here is 32 bits (i.e. the one and only IDCODE), so loading a `u32` works in my case, with the code as it is now.

However, your device has two TAPs on the scan chain, each supporting the IDCODE instruction, thus the "JTAG DR scan chain detected as 64 bits long" in your trace. Given this, my use of `load` violates the aforementioned condition from the docs.

I mistakenly thought `dr.load::<u32>()` would do the equivalent of `dr[0..32].load::<u32>()`, but I realize now that I must explicitly state that.

This made me realize that I also forgot to specify the byte order when loading; I really want to explicitly state `dr[0..32].load_le::<u32>()`.

I'll make that change and, if I can find my STM32F411 BlackPill (which also contains two TAPs daisy chained, one for the ARM DAP and the other for boundary scan), I'll use it to sanity check that I have things working with DR scan chain lengths longer than 32 bits.

cstrahan commented:
I pushed a commit to fix what I described in https://github.com/probe-rs/probe-rs/pull/1462#issuecomment-1642669832.

Here are the relevant logs:

```
{"timestamp":"2023-07-19T22:02:14.530108Z","level":"INFO","fields":{"message":"JTAG DR scan chain detected as 64 bits long"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":490,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-19T22:02:14.530773Z","level":"INFO","fields":{"message":"JTAG IR scan chain detected as 9 bits long"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":490,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-19T22:02:14.531047Z","level":"INFO","fields":{"message":"Found IDCODE: 0x4BA00477 (ARM Ltd)"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":282,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-19T22:02:14.531108Z","level":"INFO","fields":{"message":"Found IDCODE: 0x06433041 (STMicroelectronics)"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":282,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
{"timestamp":"2023-07-19T22:02:14.531153Z","level":"INFO","fields":{"message":"IR lengths are unambiguous: [4, 5]"},"target":"probe_rs::probe::cmsisdap","filename":"probe-rs/src/probe/cmsisdap/mod.rs","line_number":390,"span":{"name":"debug_port_setup"},"spans":[{"name":"debug_port_setup"}]}
```

Cleaned up a bit:

```
JTAG DR scan chain detected as 64 bits long
JTAG IR scan chain detected as 9 bits long
Found IDCODE: 0x4BA00477 (ARM Ltd)
Found IDCODE: 0x06433041 (STMicroelectronics)
IR lengths are unambiguous: [4, 5]
```

So the issue @9names reported is now resolved.

This info can be quite useful, and given these handful of lines would only be printed when initially connecting, I decided to also change the trace level to `info`. I can change that back to `debug`, if so desired.

cstrahan commented:
The `extract_ir_lengths` and `extract_idcodes` funcs didn't really belong in the `CmsisDap` impl (in my opinion), so in https://github.com/probe-rs/probe-rs/pull/1462/commits/b72a5a2e687ca9e6a58ae706e333cba4cf97f83c I moved them to a new `jtag_util` submodule, and added some tests (inspired by @9names report).

Eventually we may want to move this out of `cmsisdap` so it can be used by the other probe modules, but I'd rather do that in a follow up PR than delay this any further.

Apart from any requested changes, I feel confident that we can merge this as it stands now.

cstrahan commented:
A quick note:  we currently hard code the `dap_index` to 0 when sending the [DAP_Transfer](https://arm-software.github.io/CMSIS_5/DAP/html/group__DAP__Transfer.html) command to the probe:

https://github.com/probe-rs/probe-rs/blob/e273c9876d31b08cf67c7fe160132967cebf0ae8/probe-rs/src/probe/cmsisdap/commands/transfer/mod.rs#L144-L152

As the code stands now, this works perfectly fine when the first device on the scan chain is the device you want to debug; this *won't* work otherwise.

Eventually, we'll want to provide the following config options to the user:

- Expected IR lengths (optional if the IR lengths can be determined unambiguously)
- The DAP index

My recommendation is to merge this as-is now (or after any nits are resolved), and then have a follow up PR that removes this limitation.

cstrahan commented:
> Stellar work, thank you so much!

Thank you for this project and your maintenance/stewardship thereof :)

> Unfortunately there is some merge conflicts right now that keep me from merging :D Would you mind resolving them? :)

Just pushed a merge commit to resolve the conflicts.

> Also, I took the liberty of opening a few followup issues around the planned tasks you mentioned - independent of whether you'll take them on or someone else.

Thanks! I'll try to participate in the follow on development, time permitting.
